### PR TITLE
global, osd/crimson: improve handling of the crimson-osd instance duplication

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by rzarzynski** [comment]
[multiple instance issue: abort crimson if resource locked due to another instance using it.](https://trello.com/c/2YAqW57s/72-multiple-instance-issue-abort-crimson-if-resource-locked-due-to-another-instance-using-it)

File: src/crimson/osd/main.cc [hunk 1]
```diff
@@ -164,7 +164,13 @@ int main(int argc, char* argv[])
         });
         local_conf().parse_config_files(conf_file_list).get();
         local_conf().parse_argv(ceph_args).get();
-        pidfile_write(local_conf()->pid_file);
+        if (auto ret = pidfile_write(local_conf()->pid_file);
```
Comment (rzarzynski) [inline src/crimson/osd/main.cc:167]: `const auto`.

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
```
write: 1 20855212~5213803
discard: 22158662~2606901
write: 2 41710424~5213803
discard: 43013874~2606901
write: 3 62565636~5213803
discard: 63869086~2606901
write: 4 83420848~5213803
discard: 84724298~2606901
write: 5 104276061~5213803
discard: 105579511~2606901
migration_commit
after commit snap: user0, block 71303168~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
after commit snap: user0, block 75497472~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
after commit snap: user0, block 79691776~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
after commit snap: user1, block 71303168~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
after commit snap: user1, block 79691776~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
after commit snap: user2, block 71303168~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
after commit snap: user2, block 79691776~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
after commit snap: user3, block 71303168~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
after commit snap: user3, block 79691776~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
after commit snap: null, block 71303168~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
after commit snap: null, block 79691776~4194304 differs
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_Migration.cc:153: Failure
Value of: src_bl.contents_equal(dst_bl)
  Actual: false
Expected: true
[  FAILED  ] TestMigration.StressLive (6345 ms)
```

tracked by https://tracker.ceph.com/issues/45694

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
```
15:53:43 - DEBUG    - cbt      - Nodes : incerta04.front.sepia.ceph.com
15:53:43 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 24.5147/29.7948  => rejected
15:53:43 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
15:53:43 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 6275.0/7627.0  => rejected
15:53:43 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 286.172/416.184  => accepted
15:53:43 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00252878/0.00208343  => rejected
15:53:43 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 24.3361/30.7752  => rejected
15:53:43 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
15:53:43 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 6230.0/7878.0  => rejected
15:53:43 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 257.979/400.17  => accepted
15:53:43 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00255697/0.0020186  => rejected
15:53:43 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 167.518/251.517  => rejected
15:53:43 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/0: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
15:53:43 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 42884.0/64388.0  => rejected
15:53:43 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 2038.53/2919.26  => accepted
15:53:43 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.000369593/0.000245246  => rejected
15:53:43 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 167.122/253.748  => rejected
15:53:43 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/1: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
15:53:43 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 42783.0/64959.0  => rejected
15:53:43 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 2414.23/2916.38  => accepted
15:53:43 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.000370556/0.000243136  => rejected
15:53:43 - WARNING  - cbt      - 12 tests failed out of 20
```

this test failed, but it proves https://github.com/ceph/ceph-build/pull/1576 .

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by rzarzynski** [comment]
```
09:25:10 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 29.8286/30.6839  => accepted
09:25:10 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
09:25:10 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 7636.0/7855.0  => accepted
09:25:10 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 142.042/280.541  => accepted
09:25:10 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00208058/0.00201177  => accepted
09:25:10 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 29.3752/29.8126  => accepted
09:25:10 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
09:25:10 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 7520.0/7632.0  => accepted
09:25:10 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 97.3721/286.658  => accepted
09:25:10 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00212226/0.00207735  => accepted
09:25:10 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 240.402/252.817  => accepted
09:25:10 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/0: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
09:25:10 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 61542.0/64721.0  => accepted
09:25:10 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 1850.56/1824.72  => accepted
09:25:10 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.0002568/0.000244263  => rejected
09:25:10 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 240.773/258.788  => rejected
09:25:10 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/1: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
09:25:10 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 61637.0/66249.0  => rejected
09:25:10 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 2022.82/1552.39  => accepted
09:25:10 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.000256431/0.000238528  => rejected
09:25:10 - WARNING  - cbt      - 4 tests failed out of 20
```

**Comment by rzarzynski** [comment]
jenkins test crimson perf

File: src/crimson/osd/main.cc [hunk 2]
```diff
@@ -164,7 +164,13 @@ int main(int argc, char* argv[])
         });
         local_conf().parse_config_files(conf_file_list).get();
         local_conf().parse_argv(ceph_args).get();
-        pidfile_write(local_conf()->pid_file);
+        if (const auto ret = pidfile_write(local_conf()->pid_file);
+            ret == -EACCES || ret == -EAGAIN) {
+          ceph_abort_msg(
+            "likely there is another crimson-osd instance with the same id");
+        } else if (ret < 0) {
+          ceph_abort_msg(fmt::format("pidfile_write failed with {}", ret));
```
Comment (ronen-fr) [inline src/crimson/osd/main.cc:172]: Can we use cpp_strerror(-ret) here?

Comment (rzarzynski) [reply src/crimson/osd/main.cc:172]: Sure, will switch in a moment.

Comment (rzarzynski) [reply src/crimson/osd/main.cc]: Switched.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-06-02_10:41:24-rados-wip-kefu-testing-2020-06-02-1412-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-06-03_13:01:39-rados-wip-kefu-testing-2020-06-02-1412-1-distro-basic-smithi/

the failures are not related.

**Comment by tchaikov** [comment]
```
14:46:25 - INFO     - cbt      - prefill/incerta02.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 30.7213/28.8815  => accepted
14:46:25 - INFO     - cbt      - prefill/incerta02.front.sepia.ceph.com/0: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
14:46:25 - INFO     - cbt      - prefill/incerta02.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 7864.0/7393.0  => accepted
14:46:25 - INFO     - cbt      - prefill/incerta02.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 314.332/289.403  => accepted
14:46:25 - INFO     - cbt      - prefill/incerta02.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00200481/0.00214704  => accepted
14:46:25 - INFO     - cbt      - prefill/incerta02.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 29.7184/30.7389  => accepted
14:46:25 - INFO     - cbt      - prefill/incerta02.front.sepia.ceph.com/1: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
14:46:25 - INFO     - cbt      - prefill/incerta02.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 7607.0/7869.0  => accepted
14:46:25 - INFO     - cbt      - prefill/incerta02.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 284.018/294.575  => accepted
14:46:25 - INFO     - cbt      - prefill/incerta02.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00208709/0.00201448  => accepted
14:46:25 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 243.257/266.367  => rejected
14:46:25 - INFO     - cbt      - rand/incerta02.front.sepia.ceph.com/0: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
14:46:25 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 62273.0/68190.0  => rejected
14:46:25 - INFO     - cbt      - rand/incerta02.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 1254.55/735.366  => accepted
14:46:25 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.000253665/0.000231696  => rejected
14:46:25 - INFO     - cbt      - rand/incerta02.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 257.363/264.736  => accepted
14:46:25 - INFO     - cbt      - rand/incerta02.front.sepia.ceph.com/1: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
14:46:25 - INFO     - cbt      - rand/incerta02.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 65884.0/67772.0  => accepted
14:46:25 - INFO     - cbt      - rand/incerta02.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 950.992/2252.34  => accepted
14:46:25 - INFO     - cbt      - rand/incerta02.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.000239981/0.000233186  => accepted
14:46:25 - WARNING  - cbt      - 3 tests failed out of 20
```

**Pull request merged**